File: lua/exercises.lua
function change(amount)
  if math.type(amount) ~= "integer" then
    error("Amount must be an integer")
  end
  if amount < 0 then
    error("Amount cannot be negative")
  end
  local counts, remaining = {}, amount
  for _, denomination in ipairs({25, 10, 5, 1}) do
    counts[denomination] = remaining // denomination
    remaining = remaining % denomination
  end
  return counts
end

function first_then_lower_case(array, predicate)
  for index,value in ipairs(array) do 
    if predicate(array[index]) == true then
      return string.lower(value);
    end
  end
  return nil;
end

function powers_generator(base, limit)
  return coroutine.create(function()
    local num = 0
    while true do
      local power = base^num
      if power > limit then
        return
      end
      coroutine.yield(power)
      num = num + 1
    end
  end)
end

function say(word)
  if word == nil then
    return ""
  end
  return function(next_word)
    if next_word == nil then
      return word
    else
      return say(word .. " " .. next_word)
    end
  end
end

function meaningful_line_count(filename)
  line_number = 0
  file = io.open(filename, 'r')
  if file == nil then
  error("No such file")
  end
  for line in file:lines() do
    local trimmed_line = line:match("^%s*(.-)%s*$")
    if trimmed_line ~= "" and not trimmed_line:match("^#") then
      line_number = line_number + 1
    end
  end
  file:close()
  return line_number
end

Quaternion = (function (class)
  class.new = function(a,b,c,d)
    return setmetatable({a = a, b = b, c = c, d = d}, {
      __index = {
        coefficients = function(self)
          return {self.a, self.b, self.c, self.d}
        end,
        conjugate = function(self)
          return class.new(self.a, self.b * -1, self.c * -1,self.d* -1)
        end
      },
      __tostring = function(self)
        local total = {}
        if self.a ~= 0 or (self.b == 0 and self.c == 0 and self.d == 0) then
          table.insert(total, tostring(self.a))
        end
        if self.b ~= 0 then
          if self.b == 1 then
            table.insert(total, (#total > 0 and "+i" or "i"))
          elseif self.b == -1 then
            table.insert(total, (#total > 0 and "-i" or "-i"))
          else
            table.insert(total, (#total > 0 and (self.b > 0 and "+" or "") .. tostring(self.b) .. "i" or tostring(self.b) .. "i"))
          end
        end
        if self.c ~= 0 then
          if self.c == 1 then
            table.insert(total, (#total > 0 and "+j" or "j"))
          elseif self.c == -1 then
            table.insert(total, (#total > 0 and "-j" or "-j"))
          else
            table.insert(total, (#total > 0 and (self.c > 0 and "+" or "") .. tostring(self.c) .. "j" or tostring(self.c) .. "j"))
          end
        end
        if self.d ~= 0 then
          if self.d == 1 then
            table.insert(total, (#total > 0 and "+k" or "k"))
          elseif self.d == -1 then
            table.insert(total, (#total > 0 and "-k" or "-k"))
          else
            table.insert(total, (#total > 0 and (self.d > 0 and "+" or "") .. tostring(self.d) .. "k" or tostring(self.d) .. "k"))
          end
        end
        local result = table.concat(total)
        if result:sub(1, 1) == "+" then
          result = result:sub(2)
        end
        return result ~= "" and result or "0"
      end,
      __add = function(self, q)
        return class.new(self.a + q.a, self.b + q.b, self.c + q.c, self.d + q.d)
      end,
      __mul = function(self, q)
        return class.new(
          self.a * q.a - self.b * q.b - self.c * q.c - self.d * q.d,
          self.a * q.b + self.b * q.a + self.c * q.d - self.d * q.c,
          self.a * q.c - self.b * q.d + self.c * q.a + self.d * q.b,
          self.a * q.d + self.b * q.c - self.c * q.b + self.d * q.a
        )
      end,
      __eq = function(self, q)
        return self.a == q.a and self.b == q.b and self.c == q.c and self.d == q.d
      end
    })
  end

  return class
end)({})

File: python/exercises.py
from dataclasses import dataclass
from collections.abc import Callable
from typing import Callable, List, Optional, Generator

def change(amount: int) -> dict[int, int]:
    if not isinstance(amount, int):
        raise TypeError('Amount must be an integer')
    if amount < 0:
        raise ValueError('Amount cannot be negative')
    counts, remaining = {}, amount
    for denomination in (25, 10, 5, 1):
        counts[denomination], remaining = divmod(remaining, denomination)
    return counts


def first_then_lower_case(array: List[str], predicate: Callable[[str], bool]) -> Optional[str]:
    for i in array:
        if predicate(i):
            return i.lower()
    return None

def powers_generator(*, base: int, limit: int) -> Generator[int, None, None]:
    num = 0
    while True:
        power = base ** num
        if power > limit:
            break
        yield power
        num +=1

def say(word: Optional[str] = None) -> Callable[[Optional[str]],str]:
    if word is None:
        return ""
    def next(next_word: Optional[str] = None) -> str:
        if next_word is None:
            return word
        return say(word + " " + next_word)
    return next


def meaningful_line_count(filename: str) -> int:
    line_number = 0
    try:
        with open(filename, 'r', encoding='utf-8') as file:
            for line in file:
                stripped_line = line.strip()
                if stripped_line != "" and not stripped_line.startswith("#"):
                    line_number += 1
    except FileNotFoundError:
        raise FileNotFoundError(f"No such file: '{filename}'")
    return line_number

@dataclass(frozen=True)
class Quaternion:
    a: float
    b: float
    c: float
    d: float

    def __add__(self, q):
        return Quaternion(self.a + q.a, self.b + q.b, self.c + q.c, self.d + q.d)

    def __mul__(self, q):
        return Quaternion(
            self.a * q.a - self.b * q.b - self.c * q.c - self.d * q.d,
            self.a * q.b + self.b * q.a + self.c * q.d - self.d * q.c,
            self.a * q.c - self.b * q.d + self.c * q.a + self.d * q.b,
            self.a * q.d + self.b * q.c - self.c * q.b + self.d * q.a
        )

    def __eq__(self, q):
        return self.a == q.a and self.b == q.b and self.c == q.c and self.d == q.d

    def __str__(self):
        total = []
        if self.a != 0 or not any([self.b, self.c, self.d]):
            total.append(f"{self.a}")
        if self.b != 0:
            if self.b < 0:
                total.append(f"{self.b}i")
            else:
                total.append(f"+{self.b}i")
        if self.c != 0:
            if self.c < 0:
                total.append(f"{self.c}j")
            else:
                total.append(f"+{self.c}j")
        if self.d != 0:
            if self.d < 0:
                total.append(f"{self.d}k")
            else:
                total.append(f"+{self.d}k")
        result = ''.join(total) if total else '0'
        if result.startswith('+'):
            result = result[1:]
        ones_in_result = ["1i","1j","1k","-1i","-1j","-1k","1.0i","1.0j","1.0k","-1.0i","-1.0j","-1.0k"]
        for term in ones_in_result:
            result = result.replace(term, term[-1])
        return result

    @property
    def coefficients(self):
        return (self.a, self.b, self.c, self.d)

    @property
    def conjugate(self):
        return Quaternion(self.a, self.b * -1, self.c * -1, self.d * -1)
